openpype/lib/events.py

```diff
@@ -15,6 +16,205 @@
 class MissingEventSystem(Exception):
     pass
 
+# Python 3.4+ supports 'partialmethod'
+# TODO remove when support for Python 2.x is completely dropped
+_partial_types = [functools.partial]
+if hasattr(functools, "partialmethod"):
+    _partial_types.append(functools.partialmethod)
+_partial_types = tuple(_partial_types)
+
+
+def _get_func_info(func, is_partial):
+    """Receive function name and path.
+
+    Args:
+        func (Callable): Function or method.
+        is_partial (bool): Function is partial.
+
+    Returns:
+        tuple[str, str, Function]: Function name, path to its file and
+            real function which should be checked for reference.
+    """
+
+    if hasattr(func, "__name__"):
+        func_name = func.__name__
+    else:
+        func_name = str(func)
+
+    if is_partial:
+        func_name = "<partial {}>".format(func_name)
+
+    # Get path to file and fallback to '<unknown path>' if fails
+    # NOTE This was added because of 'partial' functions which is handled, but
+    #   who knows what else can cause this to fail?
+    try:
+        func_path = os.path.abspath(inspect.getfile(func))
+    except TypeError:
+        func_path = "<unknown path>"
+    return func_name, func_path
+
+
+class _WrappedFunc(object):
+    """Wrap function to be able to check if reference is valid.
+
+    Can create sub-wrappers for 'partial' functions.
```

Comment:
Doesn't `partial` store the function definition like a straight connection - as such, the original function couldn't get garbage collected unless the partial would have gone defunct. As such, anything it references would never have turned invalid?

Reply:
The output of this `_WrappedFunc` is that it will convert `partial` to "weak reference partial" function. Event system should store only weak references, especially in case of methods.

Reply:
Sorry, but how are partial functions themselves and thus weak references? Or what am I missing?

E.g. according to [this](https://bugs.python.org/issue34830) partial functions are weak referencable just fine?

Reply:
Do you want to know how it works, or if it works?

```python
class SomeClass:
    def __init__(self):
        self._event_system = QueuedEventSystem()

    def method_callback(self, value):
        print(value)

    def register_callbacks(self):
        def dynamic_func(value):
            print(value)

        # This will not work
        self._event_system("text", dynamic_func)
        self._event_system("text", partial(dynamic_func, "some text"))

        # This will work
        self._event_system("text", self.method_callback)
        self._event_system("text", partial(self.method_callback, "some text"))

```

Reply:
Thanks - that's likely not the exact example since you're not registering and emitting, and also the also dynamic func doesn't go out of scope so in this example would likely work originally too?

Of course calling the method callback and dynamic func directly will fail because the argument isn't passed. So that's expected behavior.


Did you mean like this?
```python
def test_event_dynamic_func_partial():

    result = []

    def dynamic_func(value):
        result.append(value)
        
    func = partial(dynamic_func, "A")
        
    event_system = QueuedEventSystem()
    event_system.add_callback(func)
    event_system.emit("test", {}, "test")
    
    assert result == ["A"]
    
    del dynamic_func
    event_system.emit("test", {}, "test")
    
    # QUESTION: Do we expect this to still work, or not?
    assert result == ["A", "A"]
    
    # Deleting the partial should not trigger the event
    # anymore since it has gone out of scope
    del func
    event_system.emit("test", {}, "test")
    assert result == ["A", "A"]
```

Reply:
```python
class SomeClass:
    def __init__(self, event_system):
        self._event_system = event_system
        self.result = []
        self.register_callbacks()

    def method_callback(self, value):
        if isinstance(value, Event):
            value = value["value"]
        self.result.append(value)

    def register_callbacks(self):
        def dynamic_func(value):
            if isinstance(value, Event):
                value = value["value"]
            self.result.append(value)

        self._event_system.add_callback("test", dynamic_func)
        self._event_system.add_callback("test", partial(dynamic_func, "A"))

        self._event_system.add_callback("test", self.method_callback)
        self._event_system.add_callback("test", partial(self.method_callback, "B"))

        self._event_system.emit("test", {"value": "V"}, "source")

event_system = QueuedEventSystem()
some_obj = SomeClass(event_system)
event_system.emit("test", {"value": "V"}, "source")
assert some_obj.result == [
    # Emitted from 'register_callback'
    # - 'dynamic_func' was still in memory
    "V", "A", "V", "B",
    # Emitted with 'obj.emit_test_event()'
    # - 'dynamic_func' is not in memory anymore
    # - 'some_obj' is still in memory so the method exists
    "V", "B",
]
```

Reply:
Ok, so that describes an intended behavior.

What's unclear to me still is why the original behavior without the nested function wrapping, etc. would fail to produce this output for this test case?

Reply:
Everything under the line can be confirmed :)

Reply:
But the partial going out of scope and thus losing its weak reference connection is still the expected behavior, yes? How else would you make a partial function go out of scope for the event system?

Reply:
Quote myself
> The output of this _WrappedFunc is that it will convert partial to "weak reference partial" function

The `partial` object is technically replaced with `_WrappedFunc`, and in that case is the partial object probably garbage collected.

Reply:
Let me try and rephrase:

- Expected behavior: If I personally don't keep a reference to the partial function I'd expect it to get removed from the registered events because it goes out of scope.

Going based on your explanations I understand that is basically the _default_ behavior for putting a `partial` in a weak reference - which as I understand is thus the default behavior even without this WrappedFunc. As such, _why_ do we need this recursive WrappedFunc behavior? What does it solve? What does it change in behavior?